tests/pipeline_with_index_test.py

```diff
@@ -0,0 +1,570 @@
+    assert pl.compute(Item((Label(int, 12),), float)) == 2.0
+
+
+def test_can_depend_on_elements_of_param_table() -> None:
+    # This is not a valid type annotation, notsure why it works with get_type_hints
+    def use_elem(x: Item((Label(int, 1),), float)) -> str:  # type: ignore[valid-type]
+        return str(x)
+
+    pl = sl.Pipeline([use_elem])
+    pl.set_param_table(sl.ParamTable(int, {float: [1.0, 2.0, 3.0]}))
+    assert pl.compute(str) == "2.0"
+
+
+def test_can_compute_series_of_param_values() -> None:
+    pl = sl.Pipeline()
+    pl.set_param_table(sl.ParamTable(int, {float: [1.0, 2.0, 3.0]}))
+    assert pl.compute(sl.Series[int, float]) == sl.Series(int, {0: 1.0, 1: 2.0, 2: 3.0})
+
+
+def test_can_compute_series_of_derived_values() -> None:
+    def process(x: float) -> str:
+        return str(x)
+
+    pl = sl.Pipeline([process])
+    pl.set_param_table(sl.ParamTable(int, {float: [1.0, 2.0, 3.0]}))
+    assert pl.compute(sl.Series[int, str]) == sl.Series(
+        int, {0: "1.0", 1: "2.0", 2: "3.0"}
+    )
+
+
+def test_creating_pipeline_with_provider_of_series_raises() -> None:
+    series = sl.Series(int, {0: 1.0, 1: 2.0, 2: 3.0})
+
+    def make_series() -> sl.Series[int, float]:
+        return series
+
+    with pytest.raises(ValueError):
+        sl.Pipeline([make_series])
+
+
+def test_creating_pipeline_with_series_param_raises() -> None:
+    series = sl.Series(int, {0: 1.0, 1: 2.0, 2: 3.0})
+
+    with pytest.raises(ValueError):
+        sl.Pipeline([], params={sl.Series[int, float]: series})
+
+
+def test_explicit_index_of_param_table_is_forwarded_correctly() -> None:
+    def process(x: float) -> int:
+        return int(x)
+
+    pl = sl.Pipeline([process])
+    pl.set_param_table(
+        sl.ParamTable(str, {float: [1.0, 2.0, 3.0]}, index=['a', 'b', 'c'])
+    )
+    assert pl.compute(sl.Series[str, int]) == sl.Series(str, {'a': 1, 'b': 2, 'c': 3})
+
+
+def test_can_gather_index() -> None:
+    Sum = NewType("Sum", float)
+    Name = NewType("Name", str)
+
+    def gather(x: sl.Series[Name, float]) -> Sum:
+        return Sum(sum(x.values()))
+
+    def make_float(x: str) -> float:
+        return float(x)
+
+    pl = sl.Pipeline([gather, make_float])
+    pl.set_param_table(sl.ParamTable(Name, {str: ["1.0", "2.0", "3.0"]}))
+    assert pl.compute(Sum) == 6.0
+
+
+def test_can_zip() -> None:
+    Sum = NewType("Sum", str)
+    Str = NewType("Str", str)
+    Run = NewType("Run", int)
+
+    def gather_zip(x: sl.Series[Run, Str], y: sl.Series[Run, int]) -> Sum:
+        z = [f'{x_}{y_}' for x_, y_ in zip(x.values(), y.values())]
+        return Sum(str(z))
+
+    def use_str(x: str) -> Str:
+        return Str(x)
+
+    pl = sl.Pipeline([gather_zip, use_str])
+    pl.set_param_table(sl.ParamTable(Run, {str: ['a', 'a', 'ccc'], int: [1, 2, 3]}))
+
+    assert pl.compute(Sum) == "['a1', 'a2', 'ccc3']"
+
+
+def test_diamond_dependency_pulls_values_from_columns_in_same_param_table() -> None:
```

Comment:
This does not look like a diamond dependency to me. It looks like a Y: Sum depends on float which depends on Param1 and Param2. But those depend on nothing.

Reply:
I meant: they both depend on the same param table and thus index.